File: Question3/Question3.cpp
#include <bits/stdc++.h>
using namespace std;

int portfolio_profit_maximisation(int maxSum, vector<int> a, vector<int> b)
{
    vector<int> a_prefixes = {0};
    vector<int> b_prefixes = {0};
    int sum = 0;
    int a_size = a.size();
    int a_prefixes_size = 1;
    for (int i = 0; i < a_size; i++)
    {
        sum += a[i];
        if (sum > maxSum)
        {
            break;
        }
        a_prefixes.push_back(sum);
        a_prefixes_size++;
    }
    sum = 0;
    int b_size = b.size();
    for (int i = 0; i < b_size; i++)
    {
        sum += b[i];
        if (sum > maxSum)
        {
            break;
        }
        b_prefixes.push_back(sum);
    }

    int result = 0;
    int j = 0;
    int i = 0;
    j = b_prefixes.size() - 1;
    for (; i < a_prefixes_size; i++)
    {
        while (j >= 0)
        {
            int temp = a_prefixes[i] + b_prefixes[j];
            int pos = i + j;
            if (temp <= maxSum)
            {
                if (pos > result)
                {
                    result = pos;
                }
                break;
            }
            j -= 1;
        }
    }
    return result;
}


int main()
{
    ios::sync_with_stdio(false);

    int n, m, x;
    cin >> n >> m >> x;

    vector<int> a(n);
    vector<int> b(m);

    for (int i = 0; i < n; ++i)
    {
        cin >> a[i];
    }

    for (int i = 0; i < m; ++i)
    {
        cin >> b[i];
    }
    int result = portfolio_profit_maximisation(x, a, b);
    cout << result;
    return 0;
}

File: Question3/Question3.py
#!/bin/python3

import math
import os
import random
import re
import sys
from typing import List


def portfolio_profit_maximisation(maxSum: int, a: List[int], b: List[int]):
    a_prefixes: List[int] = [0]
    b_prefixes: List[int] = [0]
    sum = 0
    for item in a:
        sum += item
        if sum > maxSum:
            break
        a_prefixes.append(sum)
    sum = 0
    for item in b:
        sum += item
        if sum > maxSum:
            break
        b_prefixes.append(sum)

    result = 0
    j = 0
    i = 0
    j = len(b_prefixes) - 1
    for i in range(len(a_prefixes)):
        while j >= 0:
            temp = a_prefixes[i] + b_prefixes[j]
            pos = i+j
            if temp <= maxSum:
                if pos > result:
                    result = pos
                break
            j -= 1
    return result
    
if __name__ == '__main__':
    first_multiple_input = input().rstrip().split()

    n = int(first_multiple_input[0])

    m = int(first_multiple_input[1])

    maxSum = int(first_multiple_input[2])

    a = list(map(int, input().rstrip().split()))

    b = list(map(int, input().rstrip().split()))

    result = portfolio_profit_maximisation(maxSum, a, b)

    print(result)
